File: backend_go/models/transaction.go
package models

import (
	"time"
)

type TransactionType string

const (
	Deposit  TransactionType = "deposit"
	Purchase TransactionType = "purchase"
)

type Transaction struct {
	ID                 uint      `gorm:"primaryKey"`
	UserID             uint      `gorm:"index"`
	User               BotUser   `gorm:"foreignKey:UserID"`
	OrderID            *uint
	Type               TransactionType `gorm:"not null"`
	Amount             float64         `gorm:"not null"`
	CreatedAt          time.Time       `gorm:"not null"`
	Description        string
	PaymentGateway     string  `gorm:"size:255"`
	GatewayCommission  float64
	PlatformCommission float64
	StoreBalanceDelta  float64
}

type TransactionResponse struct {
	ID                 uint            `json:"id"`
	UserID             uint            `json:"user_id"`
	OrderID            *uint           `json:"order_id"`
	Type               TransactionType `json:"type"`
	Amount             float64         `json:"amount"`
	CreatedAt          time.Time       `json:"created_at"`
	Description        string          `json:"description"`
	PaymentGateway     string          `json:"payment_gateway"`
	GatewayCommission  float64         `json:"gateway_commission"`
	PlatformCommission float64         `json:"platform_commission"`
	StoreBalanceDelta  float64         `json:"store_balance_delta"`
}

File: shared_dtos/src/transaction.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::invoice::PaymentSystem;

#[cfg_attr(feature = "sqlx", derive(sqlx::Type))]
#[cfg_attr(feature = "sqlx", sqlx(type_name = "TEXT", rename_all = "snake_case"))]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
#[cfg_attr(feature = "ts", derive(ts_rs::TS))]
#[cfg_attr(feature = "ts", ts(export, export_to = "transaction.ts"))]
#[derive(Debug, Clone, Copy, PartialEq, Deserialize, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum TransactionType {
    Deposit,
    Purchase,
    Withdrawal,
    ReferralPayout,
    ServiceCharge,
    Refund,
}

#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
#[cfg_attr(feature = "ts", derive(ts_rs::TS))]
#[cfg_attr(
    feature = "ts",
    ts(export, export_to = "transaction.ts", rename = "Transaction")
)]
#[derive(Debug, Serialize, Deserialize)]
pub struct TransactionAdminResponse {
    pub id: i64,
    pub customer_id: Option<i64>,
    pub order_id: Option<i64>,
    pub r#type: TransactionType,
    pub amount: f64,
    pub store_balance_delta: f64,
    pub platform_commission: f64,
    pub gateway_commission: f64,
    pub created_at: DateTime<Utc>,
    pub description: Option<String>,
    pub payment_gateway: Option<PaymentSystem>,
}
